File: grind75/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        best_buy_price = max(prices) + 1
        for buy, price_buy in enumerate(prices[:-1]):
            if price_buy >= best_buy_price:
                continue

            best_buy_price = price_buy
            best_sell_price = max(prices[buy:])
            profit = max(profit, best_sell_price - best_buy_price)
        return profit
        

File: grind75/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        if prices.is_empty() {
            return 0;
        }

        let mut profit = 0;
        let mut best_buy_price = prices[0];
        for &price in &prices[1..] {
            if price < best_buy_price {
                best_buy_price = price;
            } else {
                profit = profit.max(price - best_buy_price);
            }
        }

        profit
    }
}